library/alloc/src/collections/vec_deque/mod.rs

```diff
@@ -739,15 +735,13 @@ impl<T, A: Allocator> VecDeque<T, A> {
     /// ```
     #[stable(feature = "rust1", since = "1.0.0")]
     pub fn reserve(&mut self, additional: usize) {
-        let old_cap = self.cap();
-        let used_cap = self.len() + 1;
-        let new_cap = used_cap
-            .checked_add(additional)
-            .and_then(|needed_cap| needed_cap.checked_next_power_of_two())
-            .expect("capacity overflow");
+        let new_cap = self.len.checked_add(additional).expect("capacity overflow");
```

Comment:
Should this also be checking that `head + idx` will never overflow, AKA `new_cap <= isize::MAX as usize`?

I guess that thanks to #95295 it's implicitly checked for non-ZSTs, but maybe it should be checked explicitly for ZSTs, just so that the code doesn't need to worry about special-casing them elsewhere?

Perhaps something like
```rust
if T::IS_ZST && new_cap > isize::MAX as usize {
    panic!("capacity overflow");
}
```
which isn't a breaking change since in the old implementation the internal `cap` was always a power of two (so the most possible was `isize::MAX as usize + 1`) but needed a hole (so the user-visible cap could be at most `isize::MAX`).

Reply:
Is it preferable to return `isize::MAX` as opposed to `usize::MAX` in `capacity()`?. Overall it's not that much more code to handle the `ZST` cases. Also, if we restricted the `VecDeque` max-capacity to be less than the `Vec` max-capacity, that would make it so that converting a `Vec` into a `VecDeque` might have to implicitly change the capacity, and might even be fallible if the `Vec` had `> isize::MAX as usize` elements.

Reply:
Vec->VecDeque conversions already throw capacity overflow in those situations today: <https://play.rust-lang.org/?version=stable&mode=release&edition=2021&gist=810a89a9f000c8bfc527aafb275520a1>

If you're confident that the ZST handing is good, then let's leave it as you have it.